core/primitives/src/subtrie.rs

```diff
@@ -0,0 +1,287 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Child trie related struct
+
+use parity_codec::{Encode, Decode};
+use rstd::prelude::*;
+use crate::storage::well_known_keys::CHILD_STORAGE_KEY_PREFIX;
+#[cfg(feature = "std")]
+pub use impl_serde::serialize as bytes;
+
+/// keyspace type.
+pub type KeySpace = Vec<u8>;
+
+
+/// info related to parent trie.
+/// Full key of child trie storage location
+/// and size of the prefix of this location.
+pub type ParentTrie = Vec<u8>;
+
+/// temp function to keyspace data above the db level
+pub fn keyspace_in_prefix(ks: &KeySpace, prefix: &[u8], dst: &mut[u8]) {
+	assert!(dst.len() == keyspace_prefixed_expected_len(ks, prefix));
+	dst[..ks.len()].copy_from_slice(&ks);
+	dst[ks.len()..].copy_from_slice(prefix);
+}
+
+/// len of targeted prefix with keyspace
+pub fn keyspace_prefixed_expected_len(ks: &KeySpace, prefix: &[u8]) -> usize {
```

Comment:
Why is this public?

Reply:
I think my idea was to have a api without the vec allocation, but currently only the allocated one is used.
I am quite unhappy with this, but at the same time I think keyspace should be a kvdb trait parameter (to manage this at a db level and possibly allow some optimization depending on this parameter).
But my initial implementation for kvdb will be also using prefix so it does not seem that putting that in this PR would be the right time for it.